### PR TITLE
indutor: fix issue of compute index_expr range

### LINKED ISSUE #103133: [Inductor][CPU] ValueRangeError: Invalid ranges cased model crash for hf_Longformer & AllenaiLongformerBase
### 🐛 Describe the bug

 **hf_Longformer & AllenaiLongformerBase** failed tracked in https://github.com/pytorch/pytorch/issues/93531#issuecomment-1579764091

```bash
loading model: 0it [00:00, ?it/s]
loading model: 0it [00:02, ?it/s]cpu  eval  AllenaiLongformerBase               
Using FallbackKernel: aten.cumsum
[2023-06-05 09:34:47,592] torch._inductor.scheduler: [CRITICAL] Error in codegen for ComputedBuffer(name='buf7', layout=FixedLayout('cpu', torch.float32, size=[48, 4, 256, 513], stride=[525312, 131328, 513, 1]), data=Pointwise(
  'cpu',
  torch.float32,
  def inner_fn(index):
      i0, i1, i2, i3 = index
      tmp0 = ops.index_expr(i1, torch.int32)
      tmp1 = ops.index_expr(3, torch.int32)
      tmp2 = tmp0 == tmp1
      tmp3 = ops.index_expr(i3, torch.int32)
      tmp4 = ops.index_expr(256, torch.int32)
      tmp5 = tmp3 >= tmp4
      tmp6 = ops.index_expr(ModularIndexing(656384 + i3 + 513 * i2, 512, 513), torch.int64)
      tmp7 = ops.index_expr(512, torch.int64)
      tmp8 = tmp6 < tmp7
      tmp9 = ops.load(buf6, 512 * ModularIndexing(656384 + i3 + 513 * i2, 512, 513) + 262144 * ModularIndexing(656384 + i3 + 513 * i2, 262656, 3) + 786432 * ModularIndexing(656384 + i3 + 513 * i2 + 787968 * i0, 787968, 48) + ModularIndexing(i3 + 513 * i2, 1, 512))
      tmp10 = ops.masked(tmp8, tmp9, 0.0)
      tmp11 = ops.masked(tmp5, tmp10, 0.0)
      tmp12 = ops.index_expr(3, torch.int32)
      tmp13 = ops.index_expr(3, torch.int32)
      tmp14 = tmp12 < tmp13
      tmp15 = ops.index_expr(i3, torch.int32)
      tmp16 = ops.index_expr(256, torch.int32)
      tmp17 = tmp15 >= tmp16
      tmp18 = ops.index_expr(ModularIndexing(787712 + i3 + 513 * i2, 512, 513), torch.int64)
      tmp19 = ops.index_expr(512, torch.int64)
      tmp20 = tmp18 < tmp19
      tmp21 = ops.load(buf6, 512 * ModularIndexing(787712 + i3 + 513 * i2, 512, 513) + 262144 * ModularIndexing(787712 + i3 + 513 * i2, 262656, 3) + 786432 * ModularIndexing(787712 + i3 + 513 * i2 + 787968 * i0, 787968, 48) + ModularIndexing(787712 + i3 + 513 * i2, 1, 512))
      tmp22 = ops.masked(tmp20, tmp21, 0.0)
      tmp23 = ops.masked(tmp17, tmp22, 0.0)
      tmp24 = ops.constant(0, torch.float32)
      tmp25 = ops.where(tmp17, tmp23, tmp24)
      tmp26 = ops.masked(tmp14, tmp25, 0.0)
      tmp27 = ops.constant(0, torch.float32)
      tmp28 = ops.where(tmp14, tmp26, tmp27)
      tmp29 = ops.where(tmp5, tmp11, tmp28)
      tmp30 = ops.index_expr(i1, torch.int32)
      tmp31 = ops.index_expr(3, torch.int32)
      tmp32 = tmp30 < tmp31
      tmp33 = ops.index_expr(i3, torch.int32)
      tmp34 = ops.index_expr(256, torch.int32)
      tmp35 = tmp33 >= tmp34
      tmp36 = ops.index_expr(ModularIndexing(-256 + i3 + 513 * i2 + 262656 * i1 + 787968 * i0, 512, 513), torch.int64)
      tmp37 = ops.index_expr(512, torch.int64)
      tmp38 = tmp36 < tmp37
      tmp39 = ops.load(buf6, 512 * ModularIndexing(-256 + i3 + 513 * i2 + 262656 * i1 + 787968 * i0, 512, 513) + 262144 * ModularIndexing(-256 + i3 + 513 * i2 + 262656 * i1 + 787968 * i0, 262656, 3) + 786432 * ModularIndexing(-256 + i3 + 513 * i2 + 262656 * i1 + 787968 * i0, 787968, 48) + ModularIndexing(-256 + i3 + 513 * i2 + 262656 * i1 + 787968 * i0, 1, 512))
      tmp40 = ops.masked(tmp38, tmp39, 0.0)
      tmp41 = ops.masked(tmp35, tmp40, 0.0)
      tmp42 = ops.constant(0, torch.float32)
      tmp43 = ops.where(tmp35, tmp41, tmp42)
      tmp44 = ops.masked(tmp32, tmp43, 0.0)
      tmp45 = ops.constant(0, torch.float32)
      tmp46 = ops.where(tmp32, tmp44, tmp45)
      tmp47 = ops.where(tmp2, tmp29, tmp46)
      return tmp47
  ,
  ranges=[48, 4, 256, 513],
  origin_node=select_scatter,
  origins={slice_25, slice_9, slice_scatter_2, slice_scatter_1,...
))
ERROR:common:Backend dynamo failed in warmup()
Traceback (most recent call last):
  File "/workspace/pytorch/benchmarks/dynamo/common.py", line 1551, in warmup
    fn(model, example_inputs)
  File "/workspace/pytorch/torch/_dynamo/eval_frame.py", line 286, in _fn
    return fn(*args, **kwargs)
  File "benchmarks/dynamo/huggingface.py", line 509, in forward_pass
    return mod(**inputs)
  File "/workspace/pytorch/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/workspace/pytorch/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/transformers/models/longformer/modeling_longformer.py", line 1848, in forward
    outputs = self.longformer(
  File "/workspace/pytorch/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/workspace/pytorch/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/transformers/models/longformer/modeling_longformer.py", line 1750, in forward
    encoder_outputs = self.encoder(
  File "/workspace/pytorch/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/workspace/pytorch/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/transformers/models/longformer/modeling_longformer.py", line 1294, in forward
    is_global_attn = is_index_global_attn.flatten().any().item()
  File "/workspace/pytorch/torch/_dynamo/eval_frame.py", line 439, in catch_errors
    return callback(frame, cache_size, hooks, frame_state)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 527, in _convert_frame
    result = inner_convert(frame, cache_size, hooks, frame_state)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 127, in _fn
    return fn(*args, **kwargs)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 360, in _convert_frame_assert
    return _compile(
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 430, in _compile
    out_code = transform_code_object(code, transform)
  File "/workspace/pytorch/torch/_dynamo/bytecode_transformation.py", line 1000, in transform_code_object
    transformations(instructions, code_options)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 415, in transform
    tracer.run()
  File "/workspace/pytorch/torch/_dynamo/symbolic_convert.py", line 2023, in run
    super().run()
  File "/workspace/pytorch/torch/_dynamo/symbolic_convert.py", line 707, in run
    and self.step()
  File "/workspace/pytorch/torch/_dynamo/symbolic_convert.py", line 667, in step
    getattr(self, inst.opname)(inst)
  File "/workspace/pytorch/torch/_dynamo/symbolic_convert.py", line 2111, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/workspace/pytorch/torch/_dynamo/output_graph.py", line 784, in compile_subgraph
    self.compile_and_call_fx_graph(tx, pass2.graph_output_vars(), root)
  File "/opt/conda/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/workspace/pytorch/torch/_dynamo/output_graph.py", line 857, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_dynamo/output_graph.py", line 913, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/workspace/pytorch/torch/_dynamo/output_graph.py", line 909, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/workspace/pytorch/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/workspace/pytorch/torch/__init__.py", line 1536, in __call__
    return compile_fx(model_, inputs_, config_patches=self.config)
  File "/workspace/pytorch/torch/_inductor/compile_fx.py", line 802, in compile_fx
    return aot_autograd(
  File "/workspace/pytorch/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 3707, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 3233, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 2073, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 2253, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 1527, in aot_dispatch_base
    compiled_fw = compiler(fw_module, flat_args)
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_inductor/compile_fx.py", line 757, in fw_compiler_base
    return inner_compile(
  File "/workspace/pytorch/torch/_dynamo/repro/after_aot.py", line 80, in debug_wrapper
    inner_compiled_fn = compiler_fn(gm, example_inputs)
  File "/workspace/pytorch/torch/_inductor/debug.py", line 220, in inner
    return fn(*args, **kwargs)
  File "/opt/conda/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/workspace/pytorch/torch/_inductor/compile_fx.py", line 46, in newFunction
    return old_func(*args, **kwargs)
  File "/workspace/pytorch/torch/_inductor/compile_fx.py", line 311, in compile_fx_inner
    compiled_fn = graph.compile_to_fn()
  File "/workspace/pytorch/torch/_inductor/graph.py", line 897, in compile_to_fn
    return self.compile_to_module().call
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_inductor/graph.py", line 871, in compile_to_module
    code, linemap = self.codegen()
  File "/workspace/pytorch/torch/_inductor/graph.py", line 822, in codegen
    self.scheduler.codegen()
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_inductor/scheduler.py", line 1352, in codegen
    self.get_backend(device).codegen_nodes(node.get_nodes())
  File "/workspace/pytorch/torch/_inductor/codegen/cpp.py", line 2590, in codegen_nodes
    cpp_kernel_proxy.codegen_nodes(nodes)
  File "/workspace/pytorch/torch/_inductor/codegen/cpp.py", line 2498, in codegen_nodes
    tiling_factors, tiling_indices = select_tiling(vec_dtype)
  File "/workspace/pytorch/torch/_inductor/codegen/cpp.py", line 2485, in select_tiling
    run(vec_checker)
  File "/workspace/pytorch/torch/_inductor/codegen/cpp.py", line 2414, in run
    node.run(vars, reduction_vars)
  File "/workspace/pytorch/torch/_inductor/scheduler.py", line 426, in run
    self.codegen(index_vars)
  File "/workspace/pytorch/torch/_inductor/scheduler.py", line 448, in codegen
    self._body(*index_vars)
  File "/workspace/pytorch/torch/_inductor/ir.py", line 4434, in __call__
    result = self.root_block()
  File "/workspace/pytorch/torch/_inductor/ir.py", line 4543, in __call__
    return InterpreterShim(graph, submodules).run(V.get_ops_handler())
  File "/workspace/pytorch/torch/_inductor/ir.py", line 4351, in run
    return super().run(*args, **kwargs)
  File "/workspace/pytorch/torch/fx/interpreter.py", line 138, in run
    self.env[node] = self.run_node(node)
  File "/workspace/pytorch/torch/_inductor/ir.py", line 4347, in run_node
    return super().run_node(n)
  File "/workspace/pytorch/torch/fx/interpreter.py", line 195, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/workspace/pytorch/torch/fx/interpreter.py", line 312, in call_module
    return submod(*args, **kwargs)
  File "/workspace/pytorch/torch/_inductor/ir.py", line 4488, in shim
    return V.ops.masked(mask, subblock, other)
  File "/workspace/pytorch/torch/_inductor/codegen/cpp.py", line 2074, in masked
    body()
  File "/workspace/pytorch/torch/_inductor/ir.py", line 4543, in __call__
    return InterpreterShim(graph, submodules).run(V.get_ops_handler())
  File "/workspace/pytorch/torch/_inductor/ir.py", line 4351, in run
    return super().run(*args, **kwargs)
  File "/workspace/pytorch/torch/fx/interpreter.py", line 138, in run
    self.env[node] = self.run_node(node)
  File "/workspace/pytorch/torch/_inductor/ir.py", line 4347, in run_node
    return super().run_node(n)
  File "/workspace/pytorch/torch/fx/interpreter.py", line 195, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/workspace/pytorch/torch/fx/interpreter.py", line 289, in call_method
    return getattr(self_obj, target)(*args_tail, **kwargs)
  File "/workspace/pytorch/torch/_inductor/sizevars.py", line 564, in index_expr
    return self._inner.index_expr(self._simplify(index), dtype)
  File "/workspace/pytorch/torch/_inductor/codegen/cpp.py", line 2047, in index_expr
    and can_use_int32()
  File "/workspace/pytorch/torch/_inductor/codegen/cpp.py", line 2034, in can_use_int32
    expr_ranges = get_expr_range(expr, vars_ranges)
  File "/workspace/pytorch/torch/_inductor/optimize_indexing.py", line 128, in get_expr_range
    return ValueRanges(min_val, max_val)
  File "/workspace/pytorch/torch/utils/_sympy/value_ranges.py", line 74, in __init__
    raise ValueRangeError(f"Invalid ranges [{lower}:{upper}]")
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
ValueRangeError: Invalid ranges [256:0]


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True

ERROR


```

SW	| Nightly commit	| Main commit
-- | -- | --
Pytorch|[810edae](https://github.com/pytorch/pytorch/commit/810edae)|[992bffe](https://github.com/pytorch/pytorch/commit/992bffe)
Torchbench|/|[987bbec3](https://github.com/pytorch/benchmark/commit/987bbec3)
torchaudio|[894388d](https://github.com/pytorch/audio/commit/894388d)|[72d3fe0](https://github.com/pytorch/audio/commit/72d3fe0)
torchtext|[3cf1db6](https://github.com/pytorch/text/commit/3cf1db6)| [0dc07eb](https://github.com/pytorch/text/commit/0dc07eb)
torchvision|[b52841b](https://github.com/pytorch/vision/commit/b52841b)|[f36c5de](https://github.com/pytorch/vision/commit/f36c5de)
torchdata|[a754743](https://github.com/pytorch/data/commit/a754743)|[ba31745](https://github.com/pytorch/data/commit/ba31745)
dynamo_benchmarks|[174d01b](https://github.com/pytorch/pytorch/commit/174d01b)|/

### Versions

```bash
python -m torch.backends.xeon.run_cpu --node_id 0 benchmarks/dynamo/huggingface.py --performance --inference --float32 -dcpu -n50 --inductor  --no-skip --dashboard --only AllenaiLongformerBase  --cold_start_latency
python -m torch.backends.xeon.run_cpu --node_id 0 benchmarks/dynamo/torchbench.py --performance --inference --float32 -dcpu -n50 --inductor  --no-skip --dashboard --only hf_Longformer  --cold_start_latency
```

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103147

For the CPU inductor side, there has an optimization to convert ```int64``` index_expr to ```int32``` for good performance(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L2034), but for ```ModularIndexing``` exp, we replace it as division(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/optimize_indexing.py#L73, ```ModularIndexing``` doesn't have derivative) to compute derivative and then compute the expr's value range, there may meet issue which the min value may greater than the max value(```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}```).

One solution is that we don't replace ```ModularIndexing```, but it can't get the value range.
Another solution is that return ```inf``` range when the min val is great than the max val.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
I just meet one index expr which only has one ```ModularIndexing```, for a complex expr has ```ModularIndexing``` and other operatoions,  I don't think it is ok using the current method to get the value range.

File: torch/_inductor/optimize_indexing.py [hunk 1]
```diff
@@ -125,7 +125,13 @@ def indexing_div_rep(x, y):
                 for k, v in vars_ranges.items()
             },
         )
-        return ValueRanges(min_val, max_val)
+        # min_value may be greater than max_value, such as ModularIndexing(513*i2 + i3 + 262400, 512, 513),
+        # with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}.
```
Comment (jgong5) [inline torch/_inductor/optimize_indexing.py:129]: I guess a better approach is to deduce the range with the divisor. For example, we may create a new symbol with the range [0, divisor-1] suppose the divisor is constant, and put this range to the vars_ranges to calculate the min/max value with the algorithm in this function.

File: torch/_inductor/optimize_indexing.py [hunk 2]
```diff
@@ -64,15 +65,22 @@ def range_expressable_in_32_bits(range):
 
 
 def get_expr_range(expr, vars_ranges: dict):
-    free_symbols = list(expr.free_symbols)
-    if len(free_symbols) == 0:
+    symbols = list(expr.free_symbols)
+    if len(symbols) == 0:
         return ValueRanges(expr, expr)
 
+    vars_ranges_temp = vars_ranges.copy()
```
Comment (jgong5) [inline torch/_inductor/optimize_indexing.py:72]: simpler to do `vars_ranges = vars_ranges.copy()`?

**Review by jgong5 — APPROVED** [review]

**Comment by ngimel** [comment]
We should just return (0, z-1) range for ModularIndexing, and not go through derivative computation. cc @eellison

**Comment by XiaobingSuper** [comment]
> We should just return (0, z-1) range for ModularIndexing, and not go through derivative computation. cc @eellison

Yes, for a simple expr which only has ModularIndexing, it is ok just to return (0, z-1), but for a complex expr, I think we need still to use derivative and replace ModularIndexing as a variable that has a range (0, z-1). @eellison

File: torch/_inductor/optimize_indexing.py [hunk 3]
```diff
@@ -64,15 +65,22 @@ def range_expressable_in_32_bits(range):
...
+
     def replace_symbols_for_deriv(expr):
+        cnt = itertools.count()
+
         # for the purposes of finding local, minimum, maximum, assume smoothness
         def mod_indexing_rep(x, y, z):
             if z.is_constant():
-                return x / y
+                new_var = sympy_symbol("mod_index" + f"{next(cnt)}")
```
Comment (eellison) [inline torch/_inductor/optimize_indexing.py:80]: Should we check if `x / y` has a range <= `z` and return `x / y` in that case ?

**Review by eellison — COMMENTED** [review]
possible to add test ?

**Comment by ngimel** [comment]
cc @lezcano, does this overlap with #102722?

**Comment by lezcano** [comment]
Yeah, https://github.com/pytorch/pytorch/pull/102722 is strictly better than this, but this is better than the current state so I'm fine with merging this one.

FWIW, I'm going to be on PTO for the next 3 weeks, so let's merge this one now and I'll have the other one ready when I'm back.

File: torch/_inductor/optimize_indexing.py (hunk 3 above)
Comment (XiaobingSuper) [reply torch/_inductor/optimize_indexing.py:80]: I we want to return ```x/y```,  it needs to check ```z``` is positive or not, and ```x/y``` ranges are the same sign, for example, if ```x/y``` ranges are ```[-2, 2]```, ```z ``` is ```4```, we can't direct return ```x/y```. if we want to return ```x/y```, we may need to add consider many conditions, I think using z's range is ok even if the range is huge for some cases.

**Comment by XiaobingSuper** [comment]
> possible to add test ?

I have tried to get a smaller repro test case but failed. Or do I need to test ```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}``` using ```get_expr_range```.

File: test/inductor/test_cpu_repro.py [hunk 4]
```diff
@@ -387,6 +387,26 @@ def fn(x):
             (torch.randn(8),),
         )
 
+    @patch("torch.cuda.is_available", lambda: False)
```
Comment (eellison) [inline test/inductor/test_cpu_repro.py:390]: why do you need this ?

**Review by eellison — APPROVED** [review]

File: test/inductor/test_cpu_repro.py (hunk 4 above)
Comment (XiaobingSuper) [reply test/inductor/test_cpu_repro.py:390]: Remove it.

File: torch/_inductor/optimize_indexing.py (hunk 3 above)
Comment (XiaobingSuper) [reply torch/_inductor/optimize_indexing.py:80]: Yes, added.

**Pull request closed without merging**